src/dstruct/factor_graph/variable.h

```diff
@@ -34,14 +34,18 @@ namespace dd{
     long n_start_i_factors;         // id of the first factor
 
     // the values of multinomial variables are stored in an array like this
-    // [v11 v12 ... v1m v21 ... v2n ...] 
+    // [v11 v12 ... v1m v21 ... v2n ...]
     // where v1, v2 are two multinomial variables, with domain 1-m, and 1-n respectively.
     // n_start_i_tally is the start position for the variable values in the array
     long n_start_i_tally;
 
     std::unordered_map<VariableValue, int> *domain_map; // map from value to index in the domain vector
+    std::vector<VariableValue> *domain_list; // inverse of domain_map
```

Comment:
Duplicating the information in `domain_map` as another vector here is probably a bad idea as I said.  Let's remove this and put up a comment why and maybe a utility function that gives the inverse of domain_map to control future urge to do so? :)